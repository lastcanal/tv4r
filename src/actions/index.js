import { extractPosts, filterPosts } from '../helpers'
import reddit from '../helpers/reddit'

import {
  SELECT_SUBREDDIT, INVALIDATE_SUBREDDIT,
  REQUEST_POSTS, RECEIVE_POSTS, RECEIVE_POSTS_ERROR,
  SELECT_POST, NEXT_POST, PREVIOUS_POST,
  MEDIA_FALLBACK
} from '../constants'

export const selectSubreddit = subreddit => ({
  type: SELECT_SUBREDDIT,
  subreddit
})

export const invalidateSubreddit = subreddit => ({
  type: INVALIDATE_SUBREDDIT,
  subreddit
})

export const requestPosts = subreddit => ({
  type: REQUEST_POSTS,
  subreddit
})

export const receivePosts = (subreddit, json) => ({
  type: RECEIVE_POSTS,
  subreddit,
  posts: filterPosts(extractPosts(json)),
  error: null,
  receivedAt: Date.now()
})

export const receivePostsError = (subreddit, error) => ({
  type: RECEIVE_POSTS_ERROR,
  subreddit,
  error
})


export const selectPost = (post, index) => ({
  type: SELECT_POST,
  post,
  index
})

export const nextPost = (posts) => ({
  type: NEXT_POST,
  posts
})

export const previousPost = (posts) => ({
  type: PREVIOUS_POST,
  posts
})

export const mediaFallback = () => ({
  type: MEDIA_FALLBACK
})

const fetchPosts = subreddit => dispatch => {
  dispatch(requestPosts(subreddit))
  return reddit.fetchPosts(subreddit)
    .then(response => response.json())
    .then(response => dispatch(receivePosts(subreddit, response)))
    .catch(error => dispatch(receivePostsError(subreddit, error)))
}

const shouldFetchPosts = (state, subreddit) => {
  const posts = state.postsBySubreddit[subreddit]
  if (!posts) {
    return true
  }
  if (posts.isFetching) {
    return false
  }
  return posts.didInvalidate
}

export const fetchPostsIfNeeded = subreddit => (dispatch, getState) => {
  if (shouldFetchPosts(getState(), subreddit)) {
    return dispatch(fetchPosts(subreddit))
  }
}

