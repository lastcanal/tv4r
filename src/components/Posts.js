import React, { useEffect, useState } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { withStyles, useTheme } from '@material-ui/core/styles'
import GridList from '@material-ui/core/GridList'
import GridListTile from '@material-ui/core/GridListTile'
import GridListTileBar from '@material-ui/core/GridListTileBar'
import Skeleton from '@material-ui/lab/Skeleton'
import debounce from 'lodash.debounce'
import ChromeReaderModeIcon from '@material-ui/icons/ChromeReaderMode'

import { selectPost } from '../actions'
import { contrastColor } from '../helpers'
import { mediaSelector } from '../selectors'

import { THUMBNAIL_HEIGHT, THUMBNAIL_WIDTH } from '../constants'

const styles = ({ spacing, palette }) => ({
  gridList: {
    flexWrap: 'nowrap',
    transform: 'translateZ(0)',
  },
  tile: {
    opacity: 0.8,
    margin: spacing(1),
    marginBottom: spacing(2),
    boxShadow: `-1px -4px 4px -4px ${palette.primary.dark}`,
    outline: `thin double ${palette.primary.dark}`,
    '&:hover': {
      outline: `thick double ${palette.primary.dark}`,
    },
  },
  selected: {
    border: ({ themeMode }) => (
      `2px solid ${contrastColor(themeMode, palette)}`
    ),
    outline: ({ themeMode }) => (
      `thick double ${contrastColor(themeMode, palette)}`
    ),
    display: 'block',
  },
  postText: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'column',
  },
})

const classNameForTile = (selected, post, classes) => {
  if (!post?.id || !selected?.post || selected.post.id !== post.id) {
    return classes.tile
  } else {
    return `${classes.tile} ${classes.selected}`
  }
}

const style = {
  height: THUMBNAIL_HEIGHT / 2,
  width: THUMBNAIL_WIDTH,
}

const Thumbnail = ({ post }) => {
  if (post) {
    // eslint-disable-next-line camelcase
    const { thumbnail, title, over_18 } = post
    if (thumbnail === '') {
      return <ChromeReaderModeIcon fontSize="large" style={style} />
    } else if (thumbnail === 'self') {
      return <ChromeReaderModeIcon fontSize="large" style={style} />
    } else if (thumbnail === 'image') {
      return <ChromeReaderModeIcon fontSize="large" style={style} />
    } else if (thumbnail === 'default') {
      return <ChromeReaderModeIcon fontSize="large" style={style} />
    // eslint-disable-next-line camelcase
    } else if (over_18) {
      return <h2>**** NSFW ****</h2>
    } else {
      return <img src={thumbnail} alt={title} />
    }
  }

  return <Skeleton variant="rect" />
}

Thumbnail.propTypes = {
  post: PropTypes.object,
}

const Posts = ({ posts, selected, classes, dispatch }) => {
  const refs = {}

  const onSelectPost = (nextPost, index, e) => {
    e.preventDefault()
    dispatch(selectPost(nextPost, index))
  }

  const { spacing } = useTheme()

  useEffect(() => {
    if (selected?.post?.id) {
      const element = refs['post-' + selected.post.id]
      if (element) {
        const box = element.getBoundingClientRect()
        const parentBox = element.parentNode.getBoundingClientRect()
        const left = element.offsetLeft - (
          (parentBox.width / 2).toFixed() - (box.width / 2))
        element.parentNode.scrollTo({ left, behavior: 'smooth' })
      }
    }
  }, [selected])

  const calculateColumns = () => (
    window.innerWidth / (THUMBNAIL_WIDTH + spacing(2))
  )

  const [columns, setColumns] = useState(calculateColumns())

  const onResize = debounce(() => {
    setColumns(calculateColumns())
  }, 10)

  useEffect(() => {
    window.addEventListener('resize', onResize)
    return () => (window.removeEventListener('resize', onResize))
  }, [])

  const postsIter = posts.length === 0
    ? Array.from(new Array(14))
    : posts

  return (
    <div ref={node => (refs['posts-ref'] = node)}>
      <GridList
        className={classes.gridList}
        cellHeight={THUMBNAIL_HEIGHT}
        cols={columns}
      >
        {postsIter.map((post, index) => {
          const id = post ? post.id : index
          return <GridListTile
            rows={1}
            className={classNameForTile(selected, post, classes)}
            key={id}
            ref={node => (post ? (refs['post-' + post.id] = node) : '')}
            onClick={onSelectPost.bind(this, post, index)}
          >
            <Thumbnail post={post} />
            <GridListTileBar
              title={post ? post.title : <Skeleton />}
              classes={{
                root: classes.titleBar,
                title: classes.title,
              }}
            />
          </GridListTile>
        })}
      </GridList>
    </div>
  )
}

Posts.propTypes = {
  posts: PropTypes.array.isRequired,
  selected: PropTypes.object.isRequired,
  classes: PropTypes.object.isRequired,
  dispatch: PropTypes.func.isRequired,
}

const mapStateToProps = state => {
  const { dispatch, selectedSubreddit, postsBySubreddit, config } = state
  const { themeMode, showImages, showVideos } = config
  const selectedPost = postsBySubreddit.cursor || {}
  const subreddit = postsBySubreddit[selectedSubreddit]
  const posts = subreddit?.[subreddit?.scope] || []
  return {
    dispatch,
    posts: mediaSelector({ posts, showImages, showVideos }),
    selected: selectedPost,
    visible: true,
    themeMode,
  }
}

export default Posts |> withStyles(styles) |> connect(mapStateToProps)
