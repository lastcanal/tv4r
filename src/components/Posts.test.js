import GridListTile from '@material-ui/core/GridListTile';
import { Provider } from 'react-redux'

import Posts from './Posts'
import * as actions from '../actions'

describe('Posts', () => {

  const posts = [{id: 1}, {id: 2}]

  it('should render empty Posts', () => {
    const wrapper = mount(
      <Provider store={makeStore()}>
        <Posts />
      </Provider>
    )

    expect(wrapper).toMatchSnapshot()
  })


  it('should render Posts', () => {
    const store = mockStore({
      postsBySubreddit: {
        foo: {
          items: posts,
          isFetching: false,
          didInvalidate: false
        }
      },
      selectedSubreddit: 'foo',
      selectedPost: { post: posts[0] }
    })


    const wrapper = mount(
      <Provider store={store}>
        <Posts />
      </Provider>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should navigate Posts', () => {
    const dispatch = jest.fn(() => {})
    const store = mockStore({
      postsBySubreddit: {
        foo: {
          items: posts,
          isFetching: false,
          didInvalidate: false
        }
      },
      selectedSubreddit: 'foo',
      selectedPost: { post: posts[0] }
    })

    const wrapper = mount(
      <Provider store={{...store, dispatch}}>
        <Posts  />
      </Provider>
    )

    wrapper.find(GridListTile).last().simulate('click')
    wrapper.update()

    expect(dispatch).toBeCalledWith(
      actions.selectPost(posts[1], 1))
    expect(wrapper).toMatchSnapshot()
  })

})

