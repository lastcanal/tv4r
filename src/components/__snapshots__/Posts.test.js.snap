// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts should navigate Posts 1`] = `ReactWrapper {}`;

exports[`Posts should render Posts 1`] = `ReactWrapper {}`;

exports[`Posts should render empty Posts 1`] = `ReactWrapper {}`;
